user/tests/wiring/no_fixture_power_saving/power_saving_mode.cpp

```diff
@@ -0,0 +1,235 @@
+
+    return (high_pulse + low_pulse)/1000;
+}
+
+test(POWER_SAVING_00_setup) {
+#if HAL_PLATFORM_NCP
+    modem_type = cellular_modem_type();
+    if (modem_type != DEV_SARA_R510) {
+        pushMailbox(MailboxEntry().type(MailboxEntry::Type::DATA).data(skip_test_msg, sizeof(skip_test_msg) - 1));
+        skip();
+        return;
+    }
+#endif // HAL_PLATFORM_NCP
+
+    Particle.disconnect();
+    assertTrue(waitFor(Particle.disconnected, CLOUD_DISCONNECT_TIMEOUT));
+    Particle.function("fnlp1", fnLp1);
+}
+
+test(POWER_SAVING_01_particle_publish_publishes_an_event_after_low_power_active) {
+    if (modem_type != DEV_SARA_R510) {
+        skip();
+        return;
+    }
+
+    bool publishResult = false;
+    bool lowPowerResult = false;
+    int attempts = 0;
+    uint32_t period = 0;
+    const uint32_t WAIT_FOR_LOW_POWER_ACTIVE_MS = 20000;
+    const uint32_t WAIT_FOR_LOW_POWER_MEAS_MS = 10000;
+    const uint32_t LOW_POWER_ATTEMPTS_MAX = 10; // XXX: extreme cases take up to 2.5 minutes to drop into low power mode
+    do {
+        Particle.connect();
+        assertTrue(waitFor(Particle.connected, CLOUD_CONNECT_TIMEOUT));
+        delay(WAIT_FOR_LOW_POWER_ACTIVE_MS); // wait for UPSV=1 default delay of ~9.2s before modem drops into low power mode idle mode.
+
+        period = measureAvgPeriodMs(CTS1, WAIT_FOR_LOW_POWER_MEAS_MS);
+        if (period <= 1280 + 128 && period >= 1280 - 128) { // 1.28s +/- 10% if in low power mode
+            lowPowerResult = true;
+        } else {
+            lowPowerResult = false;
+        }
+
+        publishResult = Particle.publish("my_event_low_power", "event data low power", PRIVATE | WITH_ACK);
+        // Log.info("period: %lu, publishResult: %d", period, publishResult);
+    } while (!(publishResult && lowPowerResult) && attempts++ < LOW_POWER_ATTEMPTS_MAX);
+    assertMoreOrEqual(period, 1280 - 128); // 1.28s +/- 10% if in low power mode
+    assertLessOrEqual(period, 1280 + 128);
+    assertTrue(publishResult);
+}
+
+test(POWER_SAVING_02_register_function_and_connect_to_cloud) {
+    if (modem_type != DEV_SARA_R510) {
+        skip();
+        return;
+    }
+
+    bool lowPowerResult = false;
+    int attempts = 0;
+    uint32_t period = 0;
+    const uint32_t WAIT_FOR_LOW_POWER_ACTIVE_MS = 20000;
+    const uint32_t WAIT_FOR_LOW_POWER_MEAS_MS = 10000;
+    const uint32_t LOW_POWER_ATTEMPTS_MAX = 10; // XXX: extreme cases take up to 2.5 minutes to drop into low power mode
+    do {
+        Particle.connect();
+        assertTrue(waitFor(Particle.connected, CLOUD_CONNECT_TIMEOUT));
+        delay(WAIT_FOR_LOW_POWER_ACTIVE_MS); // wait for UPSV=1 default delay of ~9.2s before modem drops into low power mode idle mode.
+
+        period = measureAvgPeriodMs(CTS1, WAIT_FOR_LOW_POWER_MEAS_MS);
+        if (period <= 1280 + 128 && period >= 1280 - 128) { // 1.28s +/- 10% if in low power mode
+            lowPowerResult = true;
+        } else {
+            lowPowerResult = false;
+        }
```

Comment:
nit: it seems like you could refactor the low power check into a helper function since the same logic is done in a couple different places. Something like
```
bool enteredSleepMode() {
    <lines 125-130>
    return lowPowerResult;
}
```
or return an int if you want the exact pulse period, etc.

Reply:
That's what measureAvgPeriodMs() is for.  I thought about refactoring more, but there is an assertTrue in there and wanted to keep that visible in the test.